File: Leetcode/79.cpp
/**
 * 79. 单词搜索 回溯
 * link：https://leetcode-cn.com/problems/word-search/
 */

class Solution
{
    static constexpr int move[4][2] = {{1, 0}, {0, 1}, {-1, 0}, {0, -1}};

public:
    bool exist(vector<vector<char>> &board, string word)
    {
        int n = board.size();
        int m = board[0].size();
        vector<vector<bool>> visited(n, vector<bool>(m, false));

        for (int i = 0; i < n; i++)
        {
            for (int j = 0; j < m; j++)
            {
                if (board[i][j] == word[0])
                {
                    if (search(board, i, j, word, 0, visited))
                        return true;
                }
            }
        }
        return false;
    }

    bool search(vector<vector<char>> &board, int x, int y, string &word, int idx, vector<vector<bool>> &visited)
    {
        if (idx == word.length() - 1)
            return true;

        visited[x][y] = true;
        for (int i = 0; i < 4; i++)
        {
            int dx = x + move[i][0];
            int dy = y + move[i][1];
            if (dx < 0 || dx >= board.size() || dy < 0 || dy >= board[0].size())
                continue;
            if (visited[dx][dy] || word[idx + 1] != board[dx][dy])
                continue;
            if (search(board, dx, dy, word, idx + 1, visited))
                return true;
        }
        visited[x][y] = false;
        return false;
    }
};

File: Leetcode/79.py
# 79. 单词搜索 回溯
# link：https://leetcode-cn.com/problems/word-search/

class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        n = len(board)
        m = len(board[0])
        size = len(word)
        move = [(1, 0), (0, 1), (-1, 0), (0, -1)]
        visited = [[False for _ in range(m)] for _ in range(n)]

        def search(x: int, y: int, idx: int) -> bool:
            if idx == size - 1:
                return True

            visited[x][y] = True
            for val in move:
                dx = x + val[0]
                dy = y + val[1]
                if dx < 0 or dx >= n or dy < 0 or dy >= m:
                    continue
                if visited[dx][dy] or word[idx+1] != board[dx][dy]:
                    continue
                if search(dx, dy, idx+1):
                    return True

            visited[x][y] = False
            return False

        for i in range(n):
            for j in range(m):
                if board[i][j] == word[0]:
                    if search(i, j, 0):
                        return True

        return False
